File: cpp/median_of_two_sorted_arrays.cc
/*
 * Filename: median_of_two_sorted_arrays.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 23 Jul 2015 11:51:18 AM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
			int len = nums1.size() + nums2.size();

			if (len & 0x1) {
				return find_kth(nums1, 0, nums2, 0, len / 2 + 1);
			} else {
				return (find_kth(nums1, 0, nums2, 0, len / 2) +
						find_kth(nums1, 0, nums2, 0, len / 2 + 1)) / 2.0;
			}
		}

	private:
		double find_kth(vector<int>& nums1, int i, vector<int>& nums2, int j, int k) {
			int m = nums1.size();
			int n = nums2.size();

			if (m-i > n-j) {
				return find_kth(nums2, j, nums1, i, k);
			}

			if (i >= m) {
				return nums2[j+k-1];
			}

			if (k == 1) {
				return min(nums1[i], nums2[j]);
			}

			int idx1 = min(m, i + k/2);
			int idx2 = j + k - (idx1 - i);

			if (nums1[idx1 - 1] > nums2[idx2 - 1]) {
				return find_kth(nums1, i, nums2, idx2, k - (idx2 - j));
			} else {
				return find_kth(nums1, idx1, nums2, j, k - (idx1 - i));
			}
		}
};

File: java/MedianofTwoSortedArrays.java
/*
 * Filename: MedianofTwoSortedArrays.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Tue 03 Mar 2015 06:49:59 PM EST
 */

public class Solution {
	public double findMedianSortedArrays(int A[], int B[]) {
		int k = A.length + B.length;
		if (k % 2 == 0) {
			return (findKth(A, 0, B, 0, k/2) + findKth(A, 0, B, 0, k/2+1)) / 2.0;
		} else {
			return findKth(A, 0, B, 0, k/2+1);
		}
	}

	private double findKth(int[] A, int a, int[] B, int b, int k) {
		if (a >= A.length) {
			return B[b+k-1];
		}
		if (b >= B.length) {
			return A[a+k-1];
		}

		if (k == 1) {
			return Math.min(A[a], B[b]);
		}

		int keyA = a+k/2-1 < A.length ? A[a+k/2-1] : Integer.MAX_VALUE;
		int keyB = b+k/2-1 < B.length ? B[b+k/2-1] : Integer.MAX_VALUE; 

		if (keyA < keyB) {
			return findKth(A, a + k/2, B, b, k - k/2);
		} else {
			return findKth(A, a, B, b + k/2, k - k/2);
		}
	}
}
